### ISSUE TITLE
Add locking capabilities to the CloudVariables

### ISSUE THREAD

**Issue by brollb** [NetsBlox/NetsBlox#2009]
For non-primitive types, such as lists, we should introduce locking of the variable to support collaborative editing of a shared list variable (such as a high score list in a game).

**Comment by brollb** [comment]
Locks should probably have a max limit of a few seconds or something. Requesting the lock should probably be a blocking call and edits made without the lock should return an error.

Should it be a write-only lock or a read-write lock?

**Comment by hamidzr** [comment]
are you suggesting this to be a burden on the user?

**Comment by brollb** [comment]
Unfortunately, since we support lists, this is pretty essential unless we disallow list data. 

I am suggesting that they can first call `getWriteLock(myVariable, password)`, then `getVariable(myVariable, password)`, (some list manipulation), `setVariable(myVariable, newList, password)`, and finally `releaseWriteLock(myVariable)`.

There might be better terminology for like `freezeVariable` but I am not sure...

**Comment by brollb** [comment]
I think locks could be important for teaching concurrency in a distributed system.

**Comment by hamidzr** [comment]
I agree but are we gonna have situations where teaching about mutexes and
concurrency isn't the point when collaborating?
Hamid Zare

**Comment by brollb** [comment]
I actually think it is pretty unavoidable given the current RPCs for cloud variables (supporting lists). The good news is that it won't likely happen with few users (or if they are just writing new values independent of the earlier value). 

I say that it is unavoidable since we do not know if the value passed to `setVariable` is a modification of the previous value or is a new value entirely. If we knew it was a modification, it might be possible to do a three-way merge (or similar approach). However, we do not know if it is a modification or overwrite and the operation performed on the data may not be obvious either (so merging the changes also may not be clear).

I think that the existing RPCs will work in most cases and when users are advanced enough to run into these issues, we should be able to introduce locking. Thoughts?

**Comment by hamidzr** [comment]
Good point.  Maybe an 'update' or modify rpc could be used or a ' compare and swap' method. Using an explicit lock also seems like a good option

**Comment by brollb** [comment]
Maybe these should just be called `lockVariable` and `unlockVariable`. This might be easier to understand than specifying read vs write...

**Comment by brollb** [comment]
When a variable is locked...
- subsequent `lockVariable` calls should block on the client and be queued
- `unlockVariable` should no-op if the variable is not locked by the current user
- `setVariable` should return an error (if called by a user without the lock)
- `readVariable` should return an error (if called by a user without the lock)

Additionally, I had imagined that the locks should expire on their own naturally but maybe this actually should not be the case... This is hard to say... We probably want an expiration of some sort since the lock will be held by the unique client ID and not the role/room. However, the duration is still hard to say. Maybe something like 5 minutes would be ideal - it would be long enough to ensure any computation would likely have plenty of time to complete but not so long that an error (refreshing the page after acquiring a lock) would cost the student a whole class period.

**Comment by hamidzr** [comment]
> since the lock will be held by the unique client ID and not the role/room. 

Why is the lock held by socketid and not public user id

**Comment by brollb** [comment]
yeah, the lock should be held by a single running instance - not a named client (which may have multiple users occupying it).
